src/eepromfs.c

```diff
@@ -0,0 +1,933 @@
+    for ( size_t i = 1; i < eepfs_files_count; ++i )
+    {
+        if ( strcmp(normalized_path, eepfs_files[i].path) == 0 )
+        {
+            /* A file exists matching that path */
+            return i;
+        }
+    }
+
+    /* File does not exist */
+    return EEPFS_ENOFILE;
+}
+
+/**
+ * @brief Resolves a file handle to a file descriptor.
+ * 
+ * File descriptors can be used for reading or writing.
+ * 
+ * Getting a file repeatedly will always return the same pointer.
+ * This means that the cursor is shared across all usage of the file.
+ * Be sure to reset the cursor when you are done with it!
+ * 
+ * Negative handles indicate an upstream error occurred.
+ * The zero value will never be used as a file handle, even
+ * though it does map to an entry in the `eepfs_files` table.
+ * The signature entry is not intended to be used as a file.
+ * 
+ * @param[in] handle
+ *            The file handle of an entry in the `eepfs_files` table
+ * 
+ * @return A pointer to the file descriptor or NULL if the handle is invalid
+ */
+static eepfs_file_t * const eepfs_get_file(const int handle)
+{
+    /* Check if the handle appears to be valid */
+    if ( handle > 0 && handle < eepfs_files_count )
+    {
+        return &eepfs_files[handle];
+    }
+
+    return NULL;
+}
+
+/**
+ * @brief Reads a specific amount of data from a file starting from the cursor.
+ *
+ * @param[in]  path
+ *             Path of file in EEPROM filesystem to read from
+ * @param[out] dest
+ *             Buffer to read into
+ * @param[in]  len
+ *             Number of bytes to read
+ *
+ * @return The actual number of bytes read or a negative value on failure.
+ */
+static int eepfs_cursor_read(eepfs_file_t * const file, void * const dest, const size_t len)
+{
+    if ( file == NULL )
+    {
+        return EEPFS_EBADHANDLE;
+    }
+
+    if ( dest == NULL ) 
+    {
+        return EEPFS_EBADINPUT;
+    }
+
+    size_t cursor = file->cursor;
+    size_t bytes_to_read = len;
+
+    /* Bounds check to make sure it doesn't read past the end */
+    if ( cursor + bytes_to_read >= file->max_cursor )
+    {
+        /* It will, let's shorten it */
+        bytes_to_read = file->max_cursor - cursor;
+    }
+
+    if ( bytes_to_read <= 0 )
+    {
+        /* No bytes to read; bail early */
+        return 0;
+    }
+
+    /* Calculate the starting point to read out of EEPROM from */
+    size_t current_block = file->start_block + (cursor / EEPROM_BLOCK_SIZE);
+    size_t block_byte_offset = cursor % EEPROM_BLOCK_SIZE;
+    int bytes_read = 0;
+
+    uint8_t eeprom_buf[EEPROM_BLOCK_SIZE];
+    uint8_t * ptr = dest;
+
+    do {
+        eeprom_read(current_block, eeprom_buf);
+        /* Fill the buffer with the data from the current block */
+        while ( bytes_to_read > 0 && block_byte_offset < EEPROM_BLOCK_SIZE )
+        {
+            *(ptr++) = eeprom_buf[block_byte_offset];
+            bytes_read++;
```

Comment:
As `eepfs_file_t`'s contents are writable now, we may just mutate it here to track position. We already do that at the end so that might be easier to read.